File: backend/app/models/subscription.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models.plan import Plan


class BillingCycle(str, enum.Enum):
    monthly = "monthly"
    yearly = "yearly"


class StoreType(str, enum.Enum):
    apple = "apple"
    google = "google"
    stripe = "stripe"


class UserSubscription(Base):
    __tablename__ = "user_subscriptions"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"))
    plan_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("plans.id"))
    billing_cycle: Mapped[str] = mapped_column(String(20), default="monthly")
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )
    expires_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    store_transaction_id: Mapped[str | None] = mapped_column(
        String(255), nullable=True
    )
    store_type: Mapped[str | None] = mapped_column(String(20), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow
    )

    plan: Mapped[Plan] = relationship(lazy="joined")

File: mobile/lib/models/subscription.dart
class Subscription {
  final String id;
  final String userId;
  final String planId;
  final String billingCycle;
  final String startedAt;
  final String? expiresAt;
  final bool isActive;

  const Subscription({
    required this.id,
    required this.userId,
    required this.planId,
    required this.billingCycle,
    required this.startedAt,
    this.expiresAt,
    this.isActive = true,
  });

  factory Subscription.fromJson(Map<String, dynamic> json) {
    return Subscription(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      planId: json['plan_id'] as String,
      billingCycle: json['billing_cycle'] as String,
      startedAt: json['started_at'] as String,
      expiresAt: json['expires_at'] as String?,
      isActive: json['is_active'] as bool? ?? true,
    );
  }
}
